[6.2][concurrency] Make optimize hop to executor more conservative for 6.2 around caller isolation inheriting functions.

gottesmm opened the pull request:
Explanation: Stops elimination of hop_to_executor in a few cases.

Specifically for 6.2, we are making optimize hop to executor more conservative around caller isolation inheriting functions. This means that we are:

1. No longer treating calls to caller isolation inheriting functions as having a hop in their prologue. In terms of this pass, it means that when determining dead hop to executors, we no longer think that a caller isolation inheriting function means that an earlier hop to executor is not required.

2. Treating returns from caller isolation inheriting callees as requiring a hop. The reason why we are doing this is that we can no longer assume that our caller will hop after we return.

Post 6.2, there are three main changes we are going to make:

* Forward Dataflow

Caller isolation inheriting functions will no longer be treated as suspension points meaning that we will be able to propagate hops over them and can assume that we know the actor that we are on when we enter the function. Practically this means that trees of calls that involve just nonisolated(nonsending) async functions will avoid /all/ hop to executor calls since we will be able to eliminate all of them since the dataflow will just propagate forward from the entrance that we are already on the actor.

* Backwards Dataflow

A caller isolation inheriting call site will still cause preceding hop_to_executor functions to be live. This is because we need to ensure that we are on the caller isolation inheriting actor before we hit the call site. If we are already on that actor, the hop will be eliminated by the forward pass. But if the hop has not been eliminated, then the hop must be needed to return us to the appropriate actor.

We will also keep the behavior that returns from a caller isolation inheriting function are considered to keep hop to executors alive. If we were able to propagate to a hop to executor before the return inst with the forward dataflow, then we know that we are guaranteed to still be on the relevant actor. If the hop to executor is still there, then we need it to ensure that our caller can treat the caller isolation inheriting function as a non-suspension point.

Scope: Narrow. Only causes the optimizer to be conservative in a few cases
Issues: rdar://155905383
Original PRs: https://github.com/swiftlang/swift/pull/83083
Risk: Very low. Just makes optimizer more conservative.
Testing: CI
Reviewers: @rjmccall

gottesmm committed:
[sil] Add the ability to mark a function with isolation just so we can write more concurrency tests in SIL.

Specifically, we write a string out like:

sil [isolation "$REPRESENTATION OF ISOLATION"] @function : $@convention(thin) ...

The idea is that by using a string, we avoid parsing issues of the isolation and
have flexibility. I left in the way we put isolation into the comment above
functions so I did not break any tests that rely on it. I also made it so that
we only accept this with sil tests that pass in the flag
"sil-print-function-isolation-info". I am going to for the next release put in a
full real implementation of this that allows for actor isolation to become a
true first class citizen in SIL. But for now this at least lets us write tests
in the short term.

Since this is temporary and behind a flag, I did not add support for
serialization since this is just for writing textual SIL tests.

(cherry picked from commit ee3027c2ca5677d4877a779fff3fe6ff2d276e6c)

Conflicts:
	lib/SIL/Parser/ParseSIL.cpp

gottesmm committed:
[concurrency] Make optimize hop to executor more conservative for 6.2 around caller isolation inheriting functions.

Specifically for 6.2, we are making optimize hop to executor more conservative
around caller isolation inheriting functions. This means that we are:

1. No longer treating calls to caller isolation inheriting functions as having a
hop in their prologue. In terms of this pass, it means that when determining
dead hop to executors, we no longer think that a caller isolation inheriting
function means that an earlier hop to executor is not required.

2. Treating returns from caller isolation inheriting callees as requiring a
hop. The reason why we are doing this is that we can no longer assume that our
caller will hop after we return.

Post 6.2, there are three main changes we are going to make:

* Forward Dataflow

Caller isolation inheriting functions will no longer be treated as suspension
points meaning that we will be able to propagate hops over them and can assume
that we know the actor that we are on when we enter the function. Practically
this means that trees of calls that involve just nonisolated(nonsending) async
functions will avoid /all/ hop to executor calls since we will be able to
eliminate all of them since the dataflow will just propagate forward from the
entrance that we are already on the actor.

* Backwards Dataflow

A caller isolation inheriting call site will still cause preceding
hop_to_executor functions to be live. This is because we need to ensure that we
are on the caller isolation inheriting actor before we hit the call site. If we
are already on that actor, the hop will be eliminated by the forward pass. But
if the hop has not been eliminated, then the hop must be needed to return us to
the appropriate actor.

We will also keep the behavior that returns from a caller isolation inheriting
function are considered to keep hop to executors alive. If we were able to
propagate to a hop to executor before the return inst with the forward dataflow,
then we know that we are guaranteed to still be on the relevant actor. If the
hop to executor is still there, then we need it to ensure that our caller can
treat the caller isolation inheriting function as a non-suspension point.

rdar://155905383
(cherry picked from commit b3942424c813cbffa5069a13fcc0b816607c1f67)

gottesmm commented:
@swift-ci test

gottesmm commented:
@swift-ci test

rjmccall commented on lib/SILOptimizer/Mandatory/OptimizeHopToExecutor.cpp:
```diff
@@ -311,6 +315,40 @@ void OptimizeHopToExecutor::updateNeedExecutor(int &needExecutor,
+  // point for the purpose of eliminating redundant hop to executor to not make
+  // this optimization more aggressive. Post 6.2, we will stop treating caller
+  // isolation inheriting functions as suspension points, meaning this code can
+  // be deleted.
```
This comment is just wrong, right? If the ABI is going to be that caller isolation inheriting functions have to be called on the isolation's executor, then we will need to force hops correctly prior to the call. Marking it as not a suspension point and then not treating the call as requiring the executor would allow analysis to proceed over the call and then potentially eliminate the earlier hop.

At any rate, the code is doing the right thing, although it shouldn't really be checking `async` — in today's world (without caller-isolated sync functions) that's redundant, and in a future world (with caller-isolated sync functions) that's clearly wrong. But we don't need to fix that immediately.

gottesmm replied:
From the perspective of this pass, when we say that something is not a suspension point, what it means is that we know that after executing the instruction we are guaranteed to be on the same executor. So what this is explaining is that we will allow for hop_to_executor functions to propagate over nonisolated(nonsending) functions. I do think we could add a verification that nonisolated(nonsending) functions are always on the appropriate isolation.

NOTE: The way the pass is structured, we only ever propagate a single active actor. So the only way that we could propagate an incorrect hop to executor is if something previously already made a mistake and had an incorrect hop to executor onto the wrong actor without an intermediate hop to executor back to the expected executor before the caller isolation inheriting.

gottesmm replied:
In terms of the bottom up analysis, post 6.2, we are still going to have that caller isolation inheriting makes a call site require an executor. The change I was talking about in the comment is that I am going to be able to sink this check into needsExecutor once caller isolation inheriting is no longer considered a suspension point. Since today it /is/ considered a suspension point, I have to have this code here so it runs before we check for suspension points.

rjmccall replied:
Okay, as long as we're in agreement that the call does need to be considered to require the executor.

rjmccall commented on lib/SILOptimizer/Mandatory/OptimizeHopToExecutor.cpp:
```diff
@@ -311,6 +315,40 @@ void OptimizeHopToExecutor::updateNeedExecutor(int &needExecutor,
+  // executor.
+  if (auto isolation = inst->getFunction()->getActorIsolation();
+      isolation && isolation->isCallerIsolationInheriting() &&
+      isa<ReturnInst>(inst)) {
```
Please test for a `ReturnInst` first rather than doing a useless check for the current function's isolation on literally every instruction in the function.

rjmccall commented:
Oh, and you'll need to make sure that SILGen thunks actually perform hops in the right places when converting to/from caller-isolated types, because I'm pretty sure the current manually optimizes this (because of course it doesn't normally need to establish isolation just to do bridging and reabstraction).

gottesmm committed:
[micro-opt] Flip a check to improve perf.

gottesmm commented:
@swift-ci test

gottesmm committed:
[silgen] Make sure that thunks that convert to/from nonisolated(nonsending) handle hopping correctly.

Specifically:

1. When we convert a function to nonisolated(nonsending), we need to
make sure that in the thunk we hop upon return since nonisolated(nonsending)
functions are assumed to preserve the caller's isolation.

2. When we convert a function from nonisolated(nonsending), we need to
make sure that in the thunk we hop onto the actor that we are passing in as the
isolated parameter of the nonisolated(nonsending) function. This ensures that
the nonisolated(nonsending) function can assume that it is already on its
isolated parameter's actor at function entry.

rdar://155905383

rjmccall reviewed (approved):
LGTM

gottesmm commented:
@swift-ci test
